### ISSUE TITLE
emacs-lisp checker reports false positives when function is called in `rx` macro

### ISSUE THREAD

**Issue by pkryger** [flycheck/flycheck#2117]
Thank you for taking the time to report an issue and improve Flycheck.  This template is for **actual bugs you observed**.  If you have trouble setting up Flycheck, or if you have a question, please use the relevant issue template instead.

## Checklist

- [x] I have checked existing issues for potential duplicates before creating this one.
- [x] I have read the [Troubleshooting guide][].

## Bug description

When a function is called during `rx` macro expansion the function is reported as void.

For example, when using a a sample codefrom Info node `(elisp) Extending Rx`:
```emacs-lisp
(defun n-tuple-rx (n element)
  `(seq "<"
        (group-n 1 ,element)
        ,@(mapcar (lambda (i) `(seq ?, (group-n ,i ,element)))
                  (number-sequence 2 n))
        ">"))
(rx-define n-tuple (n element) (eval (n-tuple-rx n 'element)))

(rx (n-tuple 3 (+ (in "0-9"))))
```
`flycheck` reports the following error in `*Flycheck errors*` buffer:
```
Symbol's function definition is void: n-tuple-rx
```

## Steps to reproduce

1. Start `emacs -Q`
2. Evaluate the following:
  ```emacs-lisp
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))
(package-refresh-contents)
(package-install 'flycheck)
  ```
3. Find a new file, for example `/tmp/foo.el`
4. Enable `flycheck-mode`
5. Paste the following:
  ```emacs-lisp
;; -*- lexical-binding: t -*-

(defun n-tuple-rx (n element)
  `(seq "<"
        (group-n 1 ,element)
        ,@(mapcar (lambda (i) `(seq ?, (group-n ,i ,element)))
                  (number-sequence 2 n))
        ">"))
(rx-define n-tuple (n element) (eval (n-tuple-rx n 'element)))

(rx (n-tuple 3 (+ (in "0-9"))))
  ```
7. See error in `*Flychek errors*` buffer
8. Call `M-x elisp-byte-compile-file` -> no error is reported.

[x] I have read https://emacs.stackexchange.com/questions/28429/how-do-i-troubleshoot-emacs-problems## Expected behavior

`flycheck` should not report errors when calling functions inside `rx` macro.

## Screenshots

<img width="1463" alt="Image" src="https://github.com/user-attachments/assets/5a73b4ad-c65f-46bb-a7bf-313f7d7637a1" />

## System configuration

```
Syntax checkers for buffer foo.el in emacs-lisp-mode:

First checker to run:

  emacs-lisp
    - may enable:    yes
    - may run:       t
    - executable:    Found at /nix/store/emacs/bin/emacs
    - next checkers: emacs-lisp-checkdoc

Checkers that are compatible with this mode, but will not run until properly configured:

  emacs-lisp-checkdoc (manually disabled) enable
    - may enable: yes
    - executable: Found at /nix/store/emacs/bin/emacs

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 20250226.1541
Emacs version:    31.0.50
System:           aarch64-unknown-linux-gnu
Window system:    nil
```

Emacs configuration:
- [x] Plain Emacs / Custom configuration
- [ ] Spacemacs
- [ ] Doom Emacs
- [ ] Other shared configuration

## Additional notes

[Troubleshooting guide]:  https://www.flycheck.org/en/latest/user/troubleshooting.html

**Comment by pkryger** [comment]
Tahnks for explanation. Makes sense.
